### PR TITLE
Java parses leading annotation

### LINKED ISSUE #20026: In mixed compilation, Java parser can't handle JUnit annotation
This came up at Lightbend over at https://github.com/akka/alpakka-kafka/pull/1727

```
% cat S.scala
class S

% cat J.java
//> using scala 3.4.0
//> using dep org.junit.jupiter:junit-jupiter-api:5.10.2
@TestInstance(org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS)
class J { }

% scala-cli compile .
Compiling project (Scala 3.4.0, JVM (17))
[error] ./J.java:6:1
[error] ')' expected but eof found.
```

Next, I will see if I can construct a reproducer that doesn't involve an external library.

### PR DESCRIPTION
**Pull request by som-snytt**
Motivated by #20026

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/parsing/JavaParsers.scala [hunk 1]
```diff
@@ -1080,24 +1080,30 @@ object JavaParsers {
...
+      val buf = ListBuffer.empty[Tree]
+      var start = in.offset
+      val leadingAnnots = if (in.token == AT) annotations() else Nil
       val pkg: RefTree =
-        if (in.token == AT || in.token == PACKAGE) {
-          annotations()
+        if in.token == PACKAGE then
+          if !leadingAnnots.isEmpty then
+            //if (unit.source.file.name != "package-info.java")
+            //  syntaxError(pos, "package annotations must be in file package-info.java")
```
Comment (som-snytt) [inline compiler/src/dotty/tools/dotc/parsing/JavaParsers.scala:1090]: not sure scalac needs to enforce this javac rule if the annotations are not used

File: tests/pos/i20026/test.scala [hunk 2]
```diff
@@ -0,0 +1,6 @@
+
+object Test extends App {
+  println {
+    new JTest
```
Comment (lrytz) [inline tests/pos/i20026/test.scala:4]: (same remark as in scala 2) add reference to JFun and KTest?

File: compiler/src/dotty/tools/dotc/parsing/JavaParsers.scala (hunk 1 above)
Comment (lrytz) [reply compiler/src/dotty/tools/dotc/parsing/JavaParsers.scala:1090]: I think generally we don't add code to enforce Java rules that don't affect us in any way.

**Review by lrytz — APPROVED** [review]
LGTM!

**Comment by som-snytt** [comment]
Drafting to forward port the related annotations commits from Scala 2.

**Comment by SethTisue** [comment]
@som-snytt I've merged the Scala 2 one. Does this one need to be updated or is it ready to go?

**Comment by som-snytt** [comment]
@SethTisue I assume there is more annotations work to forward port, but I haven't had a chance (will) to look at it. Undrafting because it fixes something. I guess a solar eclipse can result in either a burst of productivity or a shadowed malaise.

**Pull request closed without merging**